server/datastore/mysql/software.go

```diff
@@ -2877,301 +2838,536 @@ last_vpp_install AS (
-		byTitleID[hs.ID] = hs
-	}
-
-	if len(titleIDs) > 0 {
-		// get the software versions installed on that host
-		const versionStmt = `
-		SELECT
-			st.id as software_title_id,
-			s.id as software_id,
-			s.version,
-			s.bundle_identifier,
-			s.source,
-			hs.last_opened_at
-		FROM
-			software s
-		INNER JOIN
-			software_titles st ON s.title_id = st.id
-		INNER JOIN
-			host_software hs ON s.id = hs.software_id AND hs.host_id = ?
-		WHERE
-			st.id IN (?)
-`
-		var installedVersions []*fleet.HostSoftwareInstalledVersion
-		stmt, args, err := sqlx.In(versionStmt, host.ID, titleIDs)
-		if err != nil {
-			return nil, nil, ctxerr.Wrap(ctx, err, "building query args to list versions")
-		}
-		if err := sqlx.SelectContext(ctx, ds.reader(ctx), &installedVersions, stmt, args...); err != nil {
-			return nil, nil, ctxerr.Wrap(ctx, err, "list software versions")
-		}
+			if opts.VulnerableOnly {
+				vulnerableJoin = `
+					INNER JOIN
+						software ON software.title_id = software_titles.id
+					INNER JOIN
+						software_cve ON software_cve.software_id = software.id
+				`
+				cveMetaJoin := `INNER JOIN cve_meta ON software_cve.cve = cve_meta.cve `
 
-		// store the installed versions with the proper software entry and collect
-		// the software ids.
-		softwareIDs := make([]uint, 0, len(installedVersions))
-		bySoftwareID := make(map[uint]*fleet.HostSoftwareInstalledVersion, len(hostSoftwareList))
-		for _, ver := range installedVersions {
-			hs := byTitleID[ver.SoftwareTitleID]
-			hs.InstalledVersions = append(hs.InstalledVersions, ver)
-			softwareIDs = append(softwareIDs, ver.SoftwareID)
-			bySoftwareID[ver.SoftwareID] = ver
-		}
+				// Only join CVE table if there are filters
+				if hasCVEFilters {
+					vulnerableJoin += cveMetaJoin
+				}
+			} else {
+				vulnerableJoin = `
+					LEFT JOIN
+						software ON software.title_id = software_titles.id
+					LEFT JOIN
+						software_cve ON software_cve.software_id = software.id
+				`
+			}
 
-		if len(softwareIDs) > 0 {
-			const cveStmt = `
-			SELECT
-				sc.software_id,
-				sc.cve
+			vppAdamStatment = `
+			UNION
+			-- SELECT for vpp apps
+			%s
 			FROM
-				software_cve sc
+				software_titles
+			LEFT JOIN
+				software_installers ON software_titles.id = software_installers.title_id
+				AND software_installers.global_or_team_id = :global_or_team_id
+			INNER JOIN
+				vpp_apps ON software_titles.id = vpp_apps.title_id AND vpp_apps.platform = :host_platform
+			INNER JOIN
+				vpp_apps_teams ON vpp_apps.adam_id = vpp_apps_teams.adam_id AND vpp_apps.platform = vpp_apps_teams.platform AND vpp_apps_teams.global_or_team_id = :global_or_team_id
+			` + vulnerableJoin + `
 			WHERE
-				sc.software_id IN (?)
-			ORDER BY
-				software_id, cve
-	`
-			type softwareCVE struct {
-				SoftwareID uint   `db:"software_id"`
-				CVE        string `db:"cve"`
+				vpp_apps_teams.adam_id IN (?) 
+				AND true
+			%s
+			` + vppOnlySelfServiceClause + `
+			-- GROUP BY for vpp apps
+			%s
```

Comment:
In theory this vpp select probably does not need to be a separate query and union. We can probably roll this up into the query above. But for simplicity it is a separate query for now.

Reply:
Not following this in general because we shouldn't be filtering for vulnerabilities on available-for-install, just on inventory, and it looks like we're filtering based on available-for-install.